### PR TITLE
Fix Immoscout24 Captcha Resolution

### PR DESCRIPTION
**Pull request by DerLeole**
Solves #577 #589 #513

Immoscout seems to have entirely moved from GeeTest to AWS WAF Captcha.
This PR implments a new solver Capmonster to deal with that fact.

My reasoning behind that:
After trying for hours to get any of the existing captcha solvers to work with the new clientside AWS WAF javascript captchas, I caved and implemented @jukoson's fix using capmonster as a solver in a way that mimics the other implementations.
It should be entirely backwards compatible

To use the new solver just modify your ENV variables or config:
```
captcha:
  capmonster:
    api_key: meow
```

Shortcomings:
- The new solver implementation only supports the aws captcha for now, but I can change that down the line.
- There is also some leftover code that in theory sniffs on background requests to get iv, context (or whatever it is now), etc from the loaded javascript. However supplying all that additional information leads to unsolvable captchas for whatever reasons and yields wrong results with 2captcha in my trials. Thats why these values are discarded in the solver itself.
- The accidentally leaked API keys included in one commit have been rotated :D 

Hope this helps!

### CODE REVIEW

**Comment by codders** [comment]
Hi @DerLeole,

First of all, thank you *so* much for taking the time to implement this. It's a real gift to have an active community on a project like flathunter, and it saves me a lot of stress and headache when people step up and make contributions.

I'll add some feedback for the review - I'll try and be clear about what I consider mandatory for merging and what's just optional. But ultimately if you don't want to implement the feedback you can also just say and I will happily tidy this up for you and get it merged (and will try and preserve your commits so that you also get the attribution).

I signed up at Capmonster and did a test locally and the code works, so that's amazing for a first contribution. Re. leaking keys, you can rebase your commits (`git rebase -i HEAD~3`) and squash them together so that the commit including the keys disappears (I know it doesn't make a difference now that you've rotated the key, but nevertheless).

I'll will test later if it works for the flathunter cloud deployment. The 2captcha implementation is anyway useless at this point, so your implementation can certainly not be worse :)

Thanks again for the contribution,

Arthur

File: flathunter/abstract_crawler.py [hunk 1]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
 
+    @backoff.on_exception(wait_gen=backoff.constant,
+                        exception=CaptchaUnsolvableError,
+                        max_tries=3)
+    def resolve_awsawf(self, driver):
+        """Resolve AWS WAF Captcha"""
+
+        # Intercept background network traffic via log sniffing
+        sleep(2)
+        logs_raw = driver.get_log("performance")
```
Comment (codders) [inline flathunter/abstract_crawler.py:207]: This is some deep magic right here. Well done for finding you're way around this.

File: flathunter/abstract_crawler.py [hunk 2]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
+        for log in filter(log_filter, logs):
+            request_id = log["params"]["requestId"]
+            resp_url = log["params"]["response"]["url"]
+            if "problem" in resp_url and "awswaf" in resp_url:
+                response = driver.execute_cdp_cmd("Network.getResponseBody", {"requestId": request_id})
+                response_json = json.loads(response["body"])
+                iv = response_json["state"]["iv"]
+                context = response_json["state"]["payload"]
+                sitekey = response_json["key"]
+
```
Comment (codders) [inline flathunter/abstract_crawler.py:227]: Nitpick - not sure why we need to double blank line here. One would be plenty

File: flathunter/abstract_crawler.py [hunk 3]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
             raise
 
+    @backoff.on_exception(wait_gen=backoff.constant,
+                        exception=CaptchaUnsolvableError,
+                        max_tries=3)
+    def resolve_awsawf(self, driver):
+        """Resolve AWS WAF Captcha"""
+
+        # Intercept background network traffic via log sniffing
+        sleep(2)
```
Comment (codders) [inline flathunter/abstract_crawler.py:206]: Generally it would be nice to avoid arbitrary 'sleep's in the code, but I appreciate that we're doing weird network magic here with an uncooperative third-party system, so for the sake of having things work I'm happy to leave this in.

File: flathunter/abstract_crawler.py [hunk 4]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
+                sitekey = response_json["key"]
+
+
+        sitekey = re.findall(
+            r"apiKey: \"(.*?)\"", driver.page_source)[0]
+
+        patternChallenge = r'src="([^"]*challenge\.js)"'
+        challenge_matches = re.findall(patternChallenge, driver.page_source)
+        for match in challenge_matches:
+            print(f'Challenge SRC Value: {match}')
```
Comment (codders) [inline flathunter/abstract_crawler.py:235]: Please replace `print`with `logger.debug` where it appears in this file.

File: flathunter/abstract_crawler.py [hunk 5]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
+        patternChallenge = r'src="([^"]*challenge\.js)"'
+        challenge_matches = re.findall(patternChallenge, driver.page_source)
+        for match in challenge_matches:
+            print(f'Challenge SRC Value: {match}')   
+            challenge = match
+
+        patternJsApi = r'src="([^"]*jsapi\.js)"'
+        jsapi_matches = re.findall(patternJsApi, driver.page_source)
+        for match in jsapi_matches:
+            print(f'JsApi SRC Value: {match}')
```
Comment (codders) [inline flathunter/abstract_crawler.py:241]: Please place `print` with `logger.debug`. Also you'll probably find for calls to the `logger` that the linter will a `%s` and a second argument to an f-string (because then it doesn't have to do the string interpolation if the log call isn't triggered).

File: flathunter/abstract_crawler.py [hunk 6]
```diff
@@ -193,6 +196,71 @@ def resolve_geetest(self, driver):
...
+            captcha = self.captcha_solver.solve_awswaf(
+                sitekey,
+                iv,
+                context,
+                challenge,
+                jsapi,
+                driver.current_url
+            )
+            old_cookie = driver.get_cookie('aws-waf-token')
+            new_cookie = old_cookie
```
Comment (codders) [inline flathunter/abstract_crawler.py:254]: I see you copied this from @jukoson's code, but I'm not sure this assignment has any effect. I think python will make `old_cookie` and `new_cookie` references to the same object. Would be worth trying to replace `old_cookie` and `new_cookie` here with just `cookie` and deleting line 254.

File: flathunter/captcha/capmonster_solver.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    CaptchaBalanceEmpty,
+    CaptchaUnsolvableError,
+    GeetestResponse,
+    AwsAwfResponse,
+    RecaptchaResponse,
+)
+
+class CapmonsterSolver(CaptchaSolver):
+    """Implementation of Captcha solver for CapMonster"""
+
```
Comment (codders) [inline flathunter/captcha/capmonster_solver.py:20]: Again not sure we need the double-empty-line here, but it's not a deal-breaker.

File: flathunter/captcha/twocaptcha_solver.py [hunk 8]
```diff
@@ -51,7 +52,7 @@ def solve_recaptcha(self, google_site_key: str, page_url: str) -> RecaptchaRespo
     def __submit_2captcha_request(self, params: Dict[str, str]) -> str:
         submit_url = "http://2captcha.com/in.php"
         submit_response = requests.post(submit_url, params=params, timeout=30)
-        logger.debug("Got response from 2captcha/in: %s", submit_response.text)
+        logger.info("Got response from 2captcha/in: %s", submit_response.text)
```
Comment (codders) [inline flathunter/captcha/twocaptcha_solver.py:55]: Do you need these changes to 2captcha?

File: flathunter/captcha/twocaptcha_solver.py [hunk 9]
```diff
@@ -66,6 +67,7 @@ def __retrieve_2captcha_result(self, captcha_id: str):
             "key": self.api_key,
             "action": "get",
             "id": captcha_id,
+            "json": 0,
```
Comment (codders) [inline flathunter/captcha/twocaptcha_solver.py:70]: What does this do?

File: flathunter/captcha/twocaptcha_solver.py [hunk 10]
```diff
@@ -87,4 +89,4 @@ def __retrieve_2captcha_result(self, captcha_id: str):
             if not retrieve_response.text.startswith("OK"):
                 raise requests.HTTPError(response=retrieve_response)
 
-            return retrieve_response.text.split("|", 1)[1]
+            return retrieve_response.text.split("|", 1)[1]
```
Comment (codders) [inline flathunter/captcha/twocaptcha_solver.py:92]: Seems like an unnecessary whitespace change - please revert this.

File: flathunter/chrome_wrapper.py [hunk 11]
```diff
@@ -59,13 +59,12 @@ def get_chrome_driver(driver_arguments):
     """Configure Chrome WebDriver"""
     logger.info('Initializing Chrome WebDriver for crawler...')
     chrome_options = uc.ChromeOptions() # pylint: disable=no-member
-    if platform == "darwin":
```
Comment (codders) [inline flathunter/chrome_wrapper.py:62]: Did you deliberately remove this? What happens if you add it back? (I'm testing on Linux, so I don't use this code path).

File: flathunter/config.py [hunk 12]
```diff
@@ -36,6 +37,7 @@ class Env:
     # Captcha setup
     FLATHUNTER_2CAPTCHA_KEY = _read_env("FLATHUNTER_2CAPTCHA_KEY")
     FLATHUNTER_IMAGETYPERZ_TOKEN = _read_env("FLATHUNTER_IMAGETYPERZ_TOKEN")
+    FLATHUNTER_CAPMONSTER_KEY = _read_env("FLATHUNTER_CAPMONSTER_KEY")
```
Comment (codders) [inline flathunter/config.py:40]: Nice! Thanks for wiring up the environment config!

File: flathunter/crawler/immobilienscout.py [hunk 13]
```diff
@@ -124,6 +124,7 @@ def get_entries_from_javascript(self):
                 logger.error(
                     "IS24 bot detection has identified our script as a bot - we've been blocked"
                 )
+                logger.info(self.get_driver_force().page_source)
```
Comment (codders) [inline flathunter/crawler/immobilienscout.py:127]: Is this `info`? Or is it enough if this is `debug`?

**Review by codders — CHANGES_REQUESTED** [review]
Thanks again for this. You've done an amazing job of following the style and layout of the code that's there - excellent work!

**Comment by codders** [comment]
Also, the Linter has a bunch of feedback. There are some `C` messages that have always been there, but the `W` issues are new and I would need them gone before we can merge the code. Likewise, `pyright` isn't happy with the types - please resolve the typing issues.

**Comment by devHaitham481** [comment]
I added the my api key after adding funds but I still get 
```
[2024/11/13 15:44:50|_common.py              |INFO    ]: Backing off resolve_amazon(...) for 0.4s (flathunter.captcha.captcha_solver.CaptchaUnsolvableError)
[2024/11/13 15:44:58|twocaptcha_solver.py    |INFO    ]: Trying to solve amazon.
[2024/11/13 15:44:58|_common.py              |INFO    ]: Backing off resolve_amazon(...) for 0.9s (flathunter.captcha.captcha_solver.CaptchaUnsolvableError)
[2024/11/13 15:45:06|twocaptcha_solver.py    |INFO    ]: Trying to solve amazon.
[2024/11/13 15:45:06|_common.py              |ERROR   ]: Giving up resolve_amazon(...) after 3 tries (flathunter.captcha.captcha_solver.CaptchaUnsolvableError)
[2024/11/13 15:45:06|hunter.py               |INFO    ]: Error while scraping url https://www.immobilienscout24.de/Suche/radius/wohnung-mieten?centerofsearchaddress=Hamburg;20459;;;;;&numberofrooms=1.5-2.0&livingspace=35.0-65.0&exclusioncriteria=projectli
sting,swapflat&pricetype=rentpermonth&geocoordinates=53.5475;9.97941;10.0&enteredFrom=result_list: the captcha was unsolvable

```

I'm not sure if it's something I'm missing or did immoscout change stuff around their end again

**Comment by codders** [comment]
This has now been merged in #634 . Thanks again for the development here - it makes a huge difference to the project to have immoscout support back.
